File: expression-language/src/main/java/io/zeebe/el/impl/FeelEvaluationResult.java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Zeebe Community License 1.0. You may not use this file
 * except in compliance with the Zeebe Community License 1.0.
 */
package io.zeebe.el.impl;

import io.zeebe.el.EvaluationResult;
import io.zeebe.el.Expression;
import io.zeebe.el.ResultType;
import java.util.List;
import java.util.Map;
import org.agrona.DirectBuffer;

public final class FeelEvaluationResult implements EvaluationResult {

  private final Expression expression;
  private final Object result;

  private final MessagePackConverter messagePackConverter = new MessagePackConverter();

  public FeelEvaluationResult(final Expression expression, final Object result) {
    this.expression = expression;
    this.result = result;
  }

  @Override
  public String getExpression() {
    return expression.getExpression();
  }

  @Override
  public boolean isFailure() {
    return false;
  }

  @Override
  public String getFailureMessage() {
    return null;
  }

  @Override
  public ResultType getType() {
    if (result == null) {
      return ResultType.NULL;

    } else if (result instanceof String) {
      return ResultType.STRING;

    } else if (result instanceof Boolean) {
      return ResultType.BOOLEAN;

    } else if (result instanceof Number) {
      return ResultType.NUMBER;

    } else if (result instanceof List) {
      return ResultType.ARRAY;

    } else if (result instanceof Map) {
      return ResultType.OBJECT;
    }

    return null;
  }

  @Override
  public DirectBuffer toBuffer() {
    return messagePackConverter.writeMessagePack(result);
  }

  @Override
  public String getString() {
    if (result instanceof String) {
      final var stringVal = (String) result;
      return stringVal;
    }
    return null;
  }

  @Override
  public Boolean getBoolean() {
    if (result instanceof Boolean) {
      final var booleanVal = (Boolean) result;
      return booleanVal;
    }
    return null;
  }

  @Override
  public Number getNumber() {
    if (result instanceof Number) {
      final var numberVal = (Number) result;
      return numberVal;
    }
    return null;
  }
}

File: expression-language/src/main/scala/io/camunda/zeebe/el/impl/feel/FeelEvaluationResult.scala
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Zeebe Community License 1.1. You may not use this file
 * except in compliance with the Zeebe Community License 1.1.
 */
package io.camunda.zeebe.el.impl.feel

import io.camunda.zeebe.el.{EvaluationResult, Expression, ResultType}
import io.camunda.zeebe.util.buffer.BufferUtil.cloneBuffer
import org.agrona.DirectBuffer
import org.camunda.feel.syntaxtree._

import java.time.{Duration, Period, ZoneId, ZonedDateTime}
import java.{lang, util}
import scala.collection.JavaConverters._

class FeelEvaluationResult(
                            expression: Expression,
                            result: Val,
                            messagePackTransformer: Val => DirectBuffer)
  extends EvaluationResult {

  override def getExpression: String = expression.getExpression

  override def isFailure: Boolean = false

  override def getFailureMessage: String = null

  override def getType: ResultType = result match {
    case ValNull => ResultType.NULL
    case _: ValBoolean => ResultType.BOOLEAN
    case _: ValNumber => ResultType.NUMBER
    case _: ValString => ResultType.STRING
    case _: ValList => ResultType.ARRAY
    case _: ValContext => ResultType.OBJECT
    case _: ValDayTimeDuration => ResultType.DURATION
    case _: ValYearMonthDuration => ResultType.PERIOD
    case _: ValDateTime => ResultType.DATE_TIME
    case _: ValLocalDateTime => ResultType.DATE_TIME
    case _ => null
  }

  override def toBuffer: DirectBuffer = messagePackTransformer(result)

  override def getString: String = result match {
    case ValString(string) => string
    case _ => null
  }

  override def getBoolean: lang.Boolean = result match {
    case ValBoolean(boolean) => boolean
    case _ => null
  }

  override def getNumber: Number = result match {
    case ValNumber(number) => number
    case _ => null
  }

  override def getDuration: Duration = result match {
    case ValDayTimeDuration(duration) => duration
    case _ => null
  }

  override def getPeriod: Period = result match {
    case ValYearMonthDuration(period) => period
    case _ => null
  }

  override def getDateTime: ZonedDateTime = result match {
    case ValDateTime(dateTime) => dateTime
    case ValLocalDateTime(dateTime) => dateTime.atZone(ZoneId.systemDefault())
    case _ => null
  }

  override def getList: util.List[DirectBuffer] = result match {
    case ValList(array) => array.map(value => cloneBuffer(messagePackTransformer(value))).asJava
    case _ => null
  }

  override def getListOfStrings: util.List[String] = result match {
    case ValList(array) => array.map {
      case ValString(string) => string
      case _ => null
    }.foldLeft(List[String]())((acc, next) => {
      if (acc == null) {
        acc
      } else if (next == null) {
        null
      } else {
        acc :+ next
      }
    }).asJava
  }
}
